### ISSUE TITLE
Refcounted binary leak with Logger in 18.0

### ISSUE THREAD

**Issue by asonge** [elixir-lang/elixir#3504]
Erlang 18.0:
```
iex> :erlang.memory(:binary)
241968
iex> require Logger
iex> for _ <- 1..100_000, do: Logger.info("test"); :ok
:ok
iex> :erlang.memory(:binary)
114548120
```

Erlang 17.0:
```
iex> :erlang.memory(:binary)
292752
iex> require Logger
iex> for _ <- 1..100_000, do: Logger.info("test"); :ok
:ok
iex> :erlang.memory(:binary)
168040
```

**Comment by josevalim** [comment]
@asonge thanks! For such short messages though... I would say they are not even refc binaries yet?

**Comment by josevalim** [comment]
If you call `:erlang.garbage_collect`, does anything change?

**Comment by asonge** [comment]
No, it does not.

Someone in IRC managed to get this to happen in vanilla erl as well, so it may not be Logger or iex.

**Comment by michalmuskala** [comment]
In erl:
```erlang
[io:format("~w~n", [X]) || X <- lists:seq(1,100000)].
erlang:memory(binary).
```

**Comment by michalmuskala** [comment]
In erl:
```
Start = erlang:memory(binary).
[io:format("~w~n", [X]) || X <- lists:seq(1,100000)].
End = erlang:memory(binary).
{Start, End}.
```
GIves `{185768,829800960}`

**Comment by arthurcolle** [comment]
Are you all on Erlang 18?

**Comment by asonge** [comment]
Yeah, this is an 18 vs 17 thing.

**Comment by StoneCypher** [comment]
aren't those just heap binaries?

http://www.erlang.org/doc/efficiency_guide/binaryhandling.html

**Comment by michalmuskala** [comment]
I don't think it's related to Logger or lists, but simply to output in console. Running:
```elixir
defmodule Test do
  def run(0), do: :ok
  def run(n) do
    IO.inspect n
    run(n - 1)
  end
end

Test.run 100000 #or other big number
```

forces uncontrollable memory growth, up to system limits, consuming all the RAM and swap space.

**Comment by fishcakez** [comment]
This is a bug in OTP unfortunately and not elixir. It would appear there is a leak in the IO sent to the shell. Since you reported this @asonge, would you like to provide a minimal erlang example and send it to erlang-bugs@erlang.org?

**Comment by dmarkow** [comment]
Looks like this was fixed upstream recently in OTP 18.0.3: http://erlang.org/pipermail/erlang-questions/2015-August/085674.html
